File: typescript/leetCode/trie/trie.ts
class TrieNode {
  public isWord: boolean;
  public word: string;
  public next: {
    [key: string]: TrieNode;
  };
  constructor() {
    this.isWord = false;
    this.next = {};
    this.word = "";
  }
}

class TrieDataStructure {
  private root: TrieNode;
  constructor() {
    this.root = new TrieNode();
  }

  public insert(word: string) {
    let curNode: TrieNode = this.root;
    for (let i: number = 0; i < word.length; i++) {
      const c: string = word[i];
      !curNode.next[c] && (curNode.next[c] = new TrieNode());
      curNode = curNode.next[c];
    }
    curNode.word = word;
    curNode.isWord = true;
  }
  public search(word: string): boolean {
    let curNode: TrieNode = this.root;
    for (let i: number = 0; i < word.length; i++) {
      curNode = curNode.next[word.charAt(i)];
      if (!curNode) return false;
    }
    return curNode.isWord;
  }

  public startsWith(prefix: string): boolean {
    let curNode = this.root;
    for (let i: number = 0; i < prefix.length; i++) {
      curNode = curNode.next[prefix.charAt(i)];
      if (!curNode) return false;
    }
    return true;
  }
}

export const buildTrie = () => {
  let trie = new TrieDataStructure();
  trie.insert("abcd");
  console.log(trie.search("b"));
};

File: cpp/code/leetCode/trie/Trie.cpp
#include <iostream>
#include <vector>
#include <unordered_map>

using namespace std;

class Node {
public:
    bool isWord;
    string word;
    unordered_map<char, Node*> next;
    Node() {
        word = "";
        isWord = false;
    }
    ~Node() { next.clear(); }
};

class Trie {
private:
    Node* root;
public:
    Trie() {
        root = new Node();
    }
    ~Trie() {
        delete root;
    }

    void insert(string word) {
        Node* curNode = root;
        for (int i = 0; i < (int)word.size(); i++) {
            char c = word[i];
            cout << c << endl;
            if (curNode->next[c] == nullptr) {
                curNode->next[c] = new Node();
            }
            curNode = curNode->next[c];
        }
        if (!curNode->isWord) {
            curNode->word = word;
            curNode->isWord = true;
        }
    }

    bool search(string word) {
        Node* curNode = root;
        for (int i = 0; i < (int)word.size(); i++) {
            char c = word[i];
            curNode = curNode->next[c];
            if (!curNode) return false;
        }
        return curNode->isWord;
    }

    bool startsWith(string prefix) {
        Node* curNode = root;
        for (int i = 0; i < prefix.size(); i++) {
            char c = prefix[i];
            curNode = curNode->next[c];
            if (!curNode) return false;
        }
        return true;
    }
};


int main() {
    Trie* trie = new Trie();
    trie->insert("app");
    trie->insert("apple");
    trie->insert("application");
    cout << trie->search("appl") << "\n";
    cout << trie->startWith("applicac") << "\n";

    return 0;
}